### PR TITLE
Database/data model

### LINKED ISSUE #7: Registration / User Data Model
## Story

**As a** Trainee/Volunteer  
**I Want** to create a new account for my role  
**So that** my identity is preserved. 

## Acceptance Criteria

1. Table created for users
2. Other related tables created (e.g. cohorts)

## Implementation Notes

1. Store trainees and volunteers in the same table.
2. I'd recommend using hard-coded constants for the role. (e.g. 0 - Trainee, 1 - Volunteer)
3. User IDs should be auto-generated.
4. Create as many tables as necessary in order not to have data duplication.
5. The output of this task is an SQL file that creates the tables.

### PR DESCRIPTION
**Pull request by Professional-Kate**
## Description

Created an SQL file that creates 6 tables (regions, cohorts, users, modules, weeks and volunteer_flags). Also included a drop table statement for each table, run the [initialize.sql](https://github.com/Professional-Kate/we-are-here-final-project/blob/database/data_model/server/initialize.sql) file to completely reset and create all the tables again.

## Related to

Resolves #7 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have carefully reviewed my own code
- [x] I have commented my code
- [x] I have updated any documentation

### CODE REVIEW

File: server/initialize.sql [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  region_id              INT NOT NULL,
+  FOREIGN KEY(region_id) REFERENCES regions(id)
+);
+
+CREATE TABLE users (
+  id                     SERIAL PRIMARY KEY,
+  first_name             VARCHAR(20) NOT NULL,
+  last_name              VARCHAR(20) NOT NULL, 
+  pass_hash              TEXT NOT NULL,
+  pass_salt              TEXT NOT NULL,
```
Comment (Professional-Kate) [inline server/initialize.sql:28]: Remove this line

File: server/initialize.sql [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+CREATE TABLE weeks (
+  id                     SERIAL PRIMARY KEY,
+  week_date              DATE NOT NULL,
+  module_id              INT NOT NULL,
+  FOREIGN KEY(module_id) REFERENCES modules(id)
+);
+
+-- I made these flags not required to make the backend code a bit easier
+CREATE TABLE volunteer_flags (
+  id                   SERIAL PRIMARY KEY,
```
Comment (Professional-Kate) [inline server/initialize.sql:55]: DEFAULT false

File: server/initialize.sql [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  CONSTRAINT check_volunteer 
+    CHECK ( (is_volunteer) OR (cohort_id IS NOT NULL))
+);             
+
+CREATE TABLE modules (
+  id   SERIAL PRIMARY KEY,
+  name VARCHAR(20) NOT NULL
+);
+
+CREATE TABLE weeks (
```
Comment (Professional-Kate) [inline server/initialize.sql:46]: start time and end time

File: server/initialize.sql [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  id                     SERIAL PRIMARY KEY,
+  week_date              DATE NOT NULL,
```
Comment (tenzyns) [inline server/initialize.sql:47]: Hi Kate, how does the weeks table sit with our data? Does it refer to JS Week-1, JS Week-2, Node Week-1? If it does, the id runs incrementally to N numbers, which can't sit with the above. Or is it a new label we are adding?

Comment (Professional-Kate) [reply server/initialize.sql:47]: Yes! So the weeks table would refer to when the weekend classes took place. Has a ``week_date`` for the day it takes place, an ``end_time`` and ``start_time`` so we know the exact time - so we can compare that to the ``volunteer_flags(clockin_time)``. The ``module_id`` is so we can add a module string to that row (JavaScript 1, HTML/CSS, React). 

The `` weeks `` table might have : 

| id | week_date | start_time | end_time | module_id
| --- | --- | --- | --- | --- |
| 1 | 12/03/2022 | 10:00 | 16:00 | 1
| 2 | 19/03/2022 | 10:00 | 16:00 | 2

The `` modules `` table might have : 

| id | name 
| --- | ---
| 1 | 'Final Project Week 2'
| 2 | 'Final Project Week 3'

Hopefully, I understood your question. Please lemme know if you have any more!

Comment (tenzyns) [reply server/initialize.sql:47]: Oh great! It's clear now. Thanks, Kate.

**Review by tenzyns — APPROVED** [review]
Looks good to me.

**Pull request merged**